go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,321 @@
+			addressSetHash = nsInfo.addressSet.GetIPv4HashName()
+		} else {
+			addressSetHash = nsInfo.addressSet.GetIPv6HashName()
+		}
+
+		for _, node := range existingNodes.Items {
+			_, stderr, err := util.RunOVNNbctl("remove", "logical_switch",
+				joinSwitch(node.Name), "acls", uuid)
+			if err != nil {
+				errList = append(errList, fmt.Errorf("failed to delete the rules for "+
+					"address_set=%s, stderr: %q (%v)", addressSetHash, stderr, err))
+			}
+
+		}
+	}
+	return errList
+}
+
+func (oc *Controller) updateEgressFirewallObjectWithRetry(egressfirewall *egressfirewallapi.EgressFirewall) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return oc.kube.UpdateEgressFirewall(egressfirewall)
+	})
+}
+
+func (ef *egressFirewall) addACLToJoinSwitch(joinSwitch, hashedAddressSetNameIPv4, hashedAddressSetNameIPv6 string) error {
+	for _, rule := range ef.egressRules {
+		var match string
+		var action string
+		if rule.access == egressfirewallapi.EgressFirewallRuleAllow {
+			action = "allow"
+		} else {
+			action = "drop"
+		}
+
+		ipAddress, _, err := net.ParseCIDR(rule.to.cidrSelector)
+		if err != nil {
+			return fmt.Errorf("error rule.to.cidrSelector %s is not a valid CIDR (%+v)", rule.to.cidrSelector, err)
+		}
+		if !utilnet.IsIPv6(ipAddress) {
+			match = fmt.Sprintf("match=\"ip4.dst == %s && ip4.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv4)
+		} else {
+			match = fmt.Sprintf("match=\"ip6.dst == %s && ip6.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv6)
+		}
+
+		if len(rule.ports) > 0 {
+
+			match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], egressGetL4Match(rule.ports))
+		}
+		uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+			"--columns=_uuid", "find", "ACL", match, "action="+action,
+			fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace))
+
+		if err != nil {
+			return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+		}
+		if uuid == "" {
+			_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
+				fmt.Sprintf("priority=%d", defaultStartPriority-rule.id),
+				fmt.Sprintf("direction=%s", fromLport), match, "action="+action,
+				fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace),
+				"--", "add", "logical_switch", joinSwitch,
+				"acls", "@acl")
+			if err != nil {
+				return fmt.Errorf("error executing create ACL command, stderr: %q, %+v", stderr, err)
+			}
+		} else {
+			_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
+			if err != nil {
+				return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
+
+			}
+		}
+
+	}
+	return nil
+}
+
+func egressGetL4Match(ports []port) string {
+	var udpString string
+	var tcpString string
+	var sctpString string
+	for _, port := range ports {
+		if port.protocol == "udp" && udpString != "udp" {
+			if port.port == 0 {
```

Comment:
Can `.port` be 0? It gets its value from https://github.com/ovn-org/ovn-kubernetes/blob/f627c0927ea8a09932da1d2e078f09698d879758/go-controller/pkg/ovn/egressfirewall.go#L75 which is always non-zero

Reply:
it is not always non-zero. since it isn't a pointer it is initialized to the default value which is 0

Reply:
Understood, my bad: I mixed up the port field with the protocol and thus thought it was required, and subsequently could never be non-zero.